s3_test.go

```diff
@@ -0,0 +1,244 @@
+package desync
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+
+	minio "github.com/minio/minio-go/v6"
+	"github.com/minio/minio-go/v6/pkg/credentials"
+	"golang.org/x/sync/errgroup"
+)
+
+type MockCredProvider struct {
+}
+
+func (p *MockCredProvider) Retrieve() (credentials.Value, error) {
+	return credentials.Value{
+		AccessKeyID:     "mainone",
+		SecretAccessKey: "thisiskeytrustmedude",
+		SessionToken:    "youdontneedtoken",
+		SignerType:      credentials.SignatureDefault,
+	}, nil
+}
+
+func (p *MockCredProvider) IsExpired() bool {
+	return false
+}
+
+func response(request *http.Request, headers http.Header, statusCode int, body string) *http.Response {
+	return &http.Response{
+		StatusCode: statusCode,
+		ProtoMajor: 1,
+		ProtoMinor: 0,
+		Request:    request,
+		Body:       io.NopCloser(strings.NewReader(body)),
+		Header:     headers,
+	}
+}
+
+func sendObject(t *testing.T, conn *net.TCPConn, request *http.Request, filePath string, sendRst bool) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			resp := response(request, http.Header{}, 404, "")
+			resp.Write(conn)
+		} else {
+			resp := response(request, http.Header{}, 500, err.Error())
+			resp.Write(conn)
+		}
+	}
```

Comment:
Shouldn't there be a `return` here or a `t.Fatal`? Without that the `file.Stat` is going to panic if file is nil (like in case of failure). Also, it's only a test, but may be worth doing a `defer file.Close()` here

Reply:
Yes, you are right, will return here and close file